OutOfSchool/OutOfSchool.WebApi/Services/ProviderService.cs

```diff
@@ -667,35 +682,105 @@ await SendNotification(checkProvider, NotificationAction.Update, statusChanged,
+    private protected async Task<Either<ErrorResponse, ActionResult>> DeleteProviderWithActionBefore(Guid id, string token, Func<Provider, Task> actionBeforeDeleting = null)
     {
-        // BUG: Possible bug with deleting provider not owned by the user itself.
-        // TODO: add unit tests to check ownership functionality
         logger.LogInformation("Deleting Provider with Id = {Id} started", id);
 
-        try
+        var entity = await providerRepository.GetWithNavigations(id).ConfigureAwait(false);
+
+        if (entity is null)
+        {
+            var message = $"There is no Provider with Id = {id}";
+            logger.LogError(message);
+            return new ErrorResponse()
+            {
+                Message = message,
+                HttpStatusCode = HttpStatusCode.NotFound,
+            };
+        }
+
+        if (!(await IsUserHasRightsAsync(currentUserService.UserId, entity)))
+        {
+            var message = $"User with userId = {currentUserService.UserId} has no rights to delete user with id = {entity.UserId}";
+            logger.LogError(message);
+            return new ErrorResponse()
+            {
+                Message = message,
+                HttpStatusCode = HttpStatusCode.Forbidden,
+            };
+        }
+
+        if (actionBeforeDeleting != null)
+        {
+            await actionBeforeDeleting(entity).ConfigureAwait(false);
+        }
+
+        await providerRepository.Delete(entity).ConfigureAwait(false);
+
+        logger.LogInformation("Provider with Id = {Id} successfully deleted", id);
+
+        await userService.Delete(entity.UserId);
+
+        logger.LogInformation("User with Id = {entity.UserId} successfully deleted", entity.UserId);
+
+        var request = new Request()
         {
-            var entity = await providerRepository.GetWithNavigations(id).ConfigureAwait(false);
+            HttpMethodType = HttpMethodType.Delete,
+            Url = new Uri(authorizationServerConfig.Authority, "account/deleteuser/" + entity.UserId),
+            Token = token,
+        };
+
+        logger.LogDebug(
+            "{HttpMethodType} Request was sent. User(id): {UserId}. Url: {Url}",
+            request.HttpMethodType,
+            currentUserService.UserId,
+            request.Url);
+
+        var response = await communicationService.SendRequest<ResponseDto>(request).ConfigureAwait(false);
 
-            if (entity is null)
+        return response
+            .FlatMap<ResponseDto>(r => r.IsSuccess
+            ? r
+            : new ErrorResponse()
+            {
+                HttpStatusCode = r.HttpStatusCode,
+                Message = r.Message,
+            })
+            .Map(r => r.Result is not null
+            ? JsonConvert.DeserializeObject<ActionResult>(r.Result.ToString())
+            : null);
+    }
+
+    private async Task<bool> IsUserHasRightsAsync(string currentUserId, Provider user)
```

Comment:
(Code duplication) This method and `IsCurrentUserHasRightsToBlockProvider` are pretty similar. Please consider redesign of both.